### PR TITLE
Performance fix for hybrid n_distinct()

### PR DESCRIPTION
**Pull request by romainfrancois**
``` r
library(dplyr)

df <- tibble(
  x=as.character(pi*sample.int(1e7,1e6,replace=TRUE)), 
  y=as.character(pi*sample.int(3e3, 1e6, replace=T))
)

system.time({
  grouped <- group_by(df, y)
  res <- summarise(grouped, n_distinct(x))
})
#>    user  system elapsed 
#>   2.568   0.033   2.606
```

<sup>Created on 2019-02-20 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

### CODE REVIEW

**Comment by romainfrancois** [comment]
closes #4202

File: inst/include/dplyr/data/NaturalDataFrame.h [hunk 1]
```diff
@@ -85,10 +85,16 @@ class NaturalDataFrame {
     return SymbolVector();
   }
 
-  static inline CharacterVector classes() {
+  static inline Rcpp::CharacterVector classes() {
+    // TODO: no need to make this vector each time
+    //       we can cache it
     return Rcpp::CharacterVector::create("tbl_df", "tbl", "data.frame");
```
Comment (krlmlr) [inline inst/include/dplyr/data/NaturalDataFrame.h]:
```suggestion
    static Rcpp::CharacterVector classes("tbl_df", "tbl", "data.frame");
    return classes;
```

Comment (romainfrancois) [reply inst/include/dplyr/data/NaturalDataFrame.h]: is this enough ? are we using these classes elsewhere too ? Is this ever released ?

**Comment by romainfrancois** [comment]
Yes I believe the first commit already fixes the problem. I just realized later that there was a bunch of stuff we no longer need.

**Comment by krlmlr** [comment]
For the class names: we should find a good place for them and use that everywhere.

**Comment by krlmlr** [comment]
Destruction is (supposedly) happening when the library is unloaded. We could double-check with a simple class that writes a log output on destruction.

**Pull request merged**